p2p/pex/pex_reactor.go

```diff
@@ -464,45 +464,113 @@ func (r *Reactor) ensurePeers() {
 	newBias := cmtmath.MinInt(out, 8)*10 + 10
 
 	toDial := make(map[p2p.ID]*p2p.NetAddress)
+	reserve := make(map[p2p.ID]*p2p.NetAddress)
 	// Try maxAttempts times to pick numToDial addresses to dial
-	maxAttempts := numToDial * 3
+	maxToDialAttempts := numToDial * 3
 
-	for i := 0; i < maxAttempts && len(toDial) < numToDial; i++ {
-		if !r.IsRunning() || !r.book.IsRunning() {
-			return
-		}
+	// Calculate reserve size as 50% of numToDial with a minimum of 10
+	reserveSize := cmtmath.MaxInt(numToDial/2, 10)
+	maxReserveAttempts := reserveSize * 3
 
-		try := r.book.PickAddress(newBias)
-		if try == nil {
-			continue
+	filter := func(ka *knownAddress) bool {
+		attempts, lastDialedTime := r.dialAttemptsInfo(ka.Addr)
+		if r.TooEarlyToDial(ka.Addr, attempts, lastDialedTime) != nil {
+			return false
+		}
+		if r.MaxAttemptsToDial(ka.Addr, attempts) != nil {
+			r.book.MarkBad(ka.Addr, defaultBanTime)
+			return false
+		}
+		if r.Switch.IsDialingOrExistingAddress(ka.Addr) {
+			return false
+		}
+		if _, selected := toDial[ka.Addr.ID]; selected {
+			return false
+		}
+		if _, selected := reserve[ka.Addr.ID]; selected {
+			return false
 		}
-		if _, selected := toDial[try.ID]; selected {
+		return true
+	}
+
+	var (
+		successCount int
+		errorCount   int
+	)
+
+	for i := 0; i < maxToDialAttempts && len(toDial) < numToDial; i++ {
+		prospectivePeer := r.book.PickAddress(newBias, filter)
+		if prospectivePeer == nil {
 			continue
 		}
-		if r.Switch.IsDialingOrExistingAddress(try) {
+		toDial[prospectivePeer.ID] = prospectivePeer
+	}
+
+	for i := 0; i < maxReserveAttempts && len(reserve) < reserveSize; i++ {
+		prospectivePeer := r.book.PickAddress(newBias, filter)
+		if prospectivePeer == nil {
 			continue
 		}
-		// TODO: consider moving some checks from toDial into here
-		// so we don't even consider dialing peers that we want to wait
-		// before dialing again, or have dialed too many times already
-		toDial[try.ID] = try
+		reserve[prospectivePeer.ID] = prospectivePeer
```

Comment:
Is this set of more addresses to dial?

Reply:
Imagine it as a subset. If a toDial address fails to dial (happens a LOT, maybe 60 percent of the time), we get a peer from the reserve pool, until the reserve pool is depleted.

I guess we could have just made toDial bigger, but it made more sense to me to imagine it this way, and hopefully we improve our addrbook heuristics more to where we can make the reserve pool smaller and smaller.

Reply:
> I guess we could have just made toDial bigger

This is the part I don't understand imho

Reply:
You don't understand as in you would have made toDial larger instead or you don't understand why toDial needs to be larger in the first place?

Reply:
you would have made toDial larger instead

Reply:
Sorry I'm still not following, you are implying we should not use reserve peers and instead just increase the size of to dial?

Reply:
Yes, I am sorry for the confusion. My question is, why not increase the size of the `trial` set?

Reply:
Yes I can do this! Just implemented it here as I was visualizing it, have no issues moving it all to toDial

Reply:
Increased the size of toDial instead of the separate sets here https://github.com/cometbft/cometbft/pull/3371/commits/6cf054c314136ca0088bca05ac266bb85b8ee1cf